ChibiOS: add CIPHERWINGF4 (STM32F405) hwdef and bootloader

jeevesh326 opened the pull request:
## Summary
Add **CIPHERWINGF4 (STM32F405)** ChibiOS hardware definition — a customized OmnibusF4-style flight controller supporting INAV, PX4, and ArduPilot firmware.

This board features optimized UART mapping, onboard OSD (MAX7456), and optional BMI270/ICM20689 IMU support.

---

## Hardware Overview

| Function | Pin | Notes |
|-----------|-----|-------|
| IMU (SPI1) | PA4 (CS), PA5 (SCK), PA6 (MISO), PA7 (MOSI) | Supports MPU6000 / BMI270 / ICM20689 |
| Baro + OSD (SPI3) | PA15 (OSD CS), PB3 (BMP280 CS), PC10–PC12 | MAX7456 OSD + BMP280 |
| I2C1 | PB8 (SCL), PB9 (SDA) | External compass / GPS |
| UART1 | PA10 (RX), PA9 (TX) | CRSF / SBUS input |
| UART2 | PA3 (RX), PA2 (TX) | Telemetry 1 |
| UART3 | PB11 (RX), PB10 (TX) | Telemetry 2 |
| UART6 | PC7 (RX), PC6 (TX) | GPS |
| USB | PA11 (DM), PA12 (DP) | OTG1 port |
| PWM Outputs | PB0–PB15, PA8 | 8 PWM outputs, DShot compatible |
| LED/Beeper | PB5 (LED), PB4 (Buzzer) | Status indicator & alarm |
| RSSI | PC1 | Analog RSSI input |
| Battery | PC2 (Voltage), PC1 (Current) | Supports external current sensor |

---

## Notes
- Based on **OmnibusF4 Pro** reference design.
- No internal compass (I2C pads provided).
- Designed and tested by **Jeevesh Vishwavijay Singh** (CipherWing Labs).
- Supports **INAV** and **PX4** unofficially, validated under ArduPilot build.

---

## Files Added
- `libraries/AP_HAL_ChibiOS/hwdef/CIPHERWINGF4/hwdef.dat`
- `libraries/AP_HAL_ChibiOS/hwdef/CIPHERWINGF4/hwdef-bl.dat`
- `libraries/AP_HAL_ChibiOS/hwdef/CIPHERWINGF4/README.md`
- `Tools/bootloaders/CIPHERWINGF4_bl.bin`

---

## Verification
- ✅ Bootloader built via `Tools/scripts/build_bootloaders.py CIPHERWINGF4`
- ✅ Firmware compiles successfully:  
Waf: Leaving directory '/home/jeevesh18/ardupilot/build/CIPHERWINGF4'
'copter' finished successfully

- ✅ Tested with INAV & PX4 builds (non-official)

---

## Future
This hardware is currently in student-stage development.  
Planned for official CipherWing Labs release after validation.
![test](https://github.com/user-attachments/assets/0b04f933-caf7-43db-981b-d6d720a9571d)
![WhatsApp Image 2025-10-29 at 11 00 34 PM](https://github.com/user-attachments/assets/0e2d6128-5817-4767-bdee-28c19d5419a6)
![WhatsApp Image 2025-10-29 at 11 00 33 PM](https://github.com/user-attachments/assets/fc828bb7-959d-4971-9a0b-01c60b89f0a8)
![back](https://github.com/user-attachments/assets/15c2ab8e-33dd-453a-b174-3790961ab8ed)
<img width="460" height="475" alt="Pasted image (3)" src="https://github.com/user-attachments/assets/fe7b1baf-e61c-447e-8508-be6eef465db2" />
<img width="460" height="475" alt="Pasted image (2)" src="https://github.com/user-attachments/assets/65e5fd09-87ab-418a-9ca0-8b849a507db4" />

jeevesh326 committed:
ChibiOS: add CIPHERWINGF4 (STM32F405) hwdef and bootloader

jeevesh326 commented:
> The provided images really do not have a good enough silkscreen for each pads function to review, however a quick scan of the readme and hwdef, indicates that SBUS is notr supported...on an F4 it needs an timer interrupt input OR a separate pad that inverts SBUS before application to a UART setup with protocol 23 (and it should have full DMA to accommodate CRSF/ELRS RC in also). FYI, the UART protocol for SBUS is OUTPUT only, not input so cant be used for SBUS RCin.

Added SBUS input support using a BSS138 inverter circuit connected to UART5_RX (PD2).
This allows proper signal inversion for SBUS input on the STM32F405 while maintaining UART1 and UART6 for telemetry or CRSF/ELRS.
	•	Updated hwdef.dat to define UART5_RX as inverted and set it as the default RC input port (HAL_RCIN_UART 5).
	•	Updated the board README to include SBUS hardware inverter details and RC input configuration.

The hardware schematic now includes an onboard BSS138 inverter between the SBUS pad and PD2.
This should resolve the SBUS compatibility concern mentioned earlier.

jeevesh326 committed:
CIPHERWINGF4: add SBUS input on UART5_RX (via BSS138 inverter)

jeevesh326 committed:
CIPHERWINGF4: restore README and document SBUS via UART5_RX inverter

jeevesh326 committed:
CIPHERWINGF4: update board_types.txt and bootloader definition

jeevesh326 commented:
✅ All checks have passed (56 successful).

I’ve addressed the previous SBUS inversion feedback by routing SBUS input through UART5_RX via BSS138 inverter, and confirmed successful ChibiOS and bootloader builds.

The board definition (CIPHERWINGF4) is stable and ready for final review.

@Hwurzburg  @peterbarker  — kindly review and approve when possible. Thank you for your guidance throughout this process!

Hwurzburg commented:
@jeevesh326 when the new images are uploaded, I can review....ping me

jeevesh326 committed:
CI: retrigger DDS geopose test flake

Hwurzburg reviewed (commented):
the new silkscreen has not BAT- pads, only BAT+
the vtx and cam rows have no labels for pads
the LED row has + and - pads...what are they? + should have its voltage labeled, and what is -? gnd?
once these are answered I can review...thanks

jeevesh326 commented:
> the new silkscreen has not BAT- pads, only BAT+
> 
> the vtx and cam rows have no labels for pads
> 
> the LED row has + and - pads...what are they? + should have its voltage labeled, and what is -? gnd?
> 
> once these are answered I can review...thanks

Hi @Hwurzburg  
I’ve updated the silkscreen again to make things clearer. The VTX and CAM pads are not labeled on the top layer due to limited space, but I’ve added clear notations on the bottom silkscreen layer for reference.
Also, I’ve updated all “+” and “–” markings to directly show “+5V” and “GND” to avoid any confusion.

Thanks for the feedback — this should make it easier to identify everything now.
![image](https://github.com/user-attachments/assets/5aa5feb9-b22e-4533-99cd-f862da084c32)

jeevesh326 commented:
you can check the user manual @Hwurzburg 
[end_manual.pdf](https://github.com/user-attachments/files/23304899/end_manual.pdf)

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/CIPHERWINGF4/hwdef.dat:
```diff
@@ -0,0 +1,181 @@
+# M6 PA8  TIM1_CH1
+# M7 PB15 TIM1_CH3N   (F405-valid replacement for TIM12_CH2)
+# M8 PB7  TIM4_CH2
+PB0  TIM3_CH3   TIM3  PWM(1) GPIO(50)
```
Omnibus F4 pro does this:
```
PB0 TIM3_CH3  TIM3 PWM(1) GPIO(50)
PB1 TIM3_CH4  TIM3 PWM(2) GPIO(51) BIDIR
PA3 TIM2_CH4  TIM2 PWM(3) GPIO(52) BIDIR
PA2 TIM2_CH3  TIM2 PWM(4) GPIO(53)
```
you may find that some of the other timers have better allocation

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/CIPHERWINGF4/hwdef.dat:
```diff
@@ -0,0 +1,181 @@
+
+# ---------- UARTs ----------
+PA10 USART1_RX USART1
+PA9  USART1_TX USART1
```
Put the standard protocol functions expected for these

Hwurzburg replied:
okay still waiting on good image to be included in readdme  before I detail review...I took it as meaning ALL the uarts need protocol defines...thanks for clarifying

andyp1per reviewed (changes requested):
What is the release plan for this? We don't want to be building firmware for something that is an experiment

Hwurzburg commented:
Closing this due to lack of responses/actions
